hazelcast/src/main/java/com/hazelcast/internal/probing/ProbingCycle.java

```diff
@@ -0,0 +1,222 @@
+
+    /**
+     * Similar to {@link #probe(Object)} just that all names become
+     * {@code <prefix>.<name>} instead of just {@code <name>}.
+     * 
+     * The method exits with the same {@link Tags} context it is called with.
+     * 
+     * @param prefix prefix to use for all names (provided without dot)
+     * @param instance a obj with fields or methods annotated with {@link Probe}
+     */
+    void probe(CharSequence prefix, Object instance);
+
+    /**
+     * Similar to {@link #probe(Object)} just that probed methods are not identified
+     * using annotations but given as a list. All methods use the specified level.
+     * No fields will be probed.
+     * 
+     * This can be used to probe beans that cannot be annotated. To prefix this
+     * {@link Tags#prefix(CharSequence)} can be used. Remember though that the
+     * prefix will remain in the context.
+     * 
+     * @param level the level to use for all probes created
+     * @param instance the obj whose methods to probe.
+     * @param methods the names of the methods to probe. Methods that do not exist
+     *        or have the wrong number of arguments (not none) are ignored.
+     */
+    void probe(ProbeLevel level, Object instance, String... methods);
+
+    void probe(CharSequence name, long value);
+
+    void probe(CharSequence name, double value);
+
+    void probe(CharSequence name, boolean value);
+
+    void probe(ProbeLevel level, CharSequence name, long value);
+
+    void probe(ProbeLevel level, CharSequence name, double value);
+
+    void probe(ProbeLevel level, CharSequence name, boolean value);
+
+    /**
+     * Used to forward metrics received from the client. This is only different from
+     * {@link #probe(CharSequence, long)} except that it will not escape the name
+     * but un-escaping escaped backslashes.
+     * 
+     * @param name a properly escaped name (as written before)
+     * @param value value as written before (double already converted to long)
+     */
+    void probeForwarded(CharSequence name, long value);
+
+    /*
+     * Describing probing context to the cycle:
+     */
+
+    /**
+     * Resets the tags used by this cycle (to empty). This provides a convenient and
+     * format independent way of tagging the probed metrics.
+     * 
+     * After context (previously the prefix) is completed using
+     * {@link Tags#tag(CharSequence, CharSequence)} the {@code probe} methods are
+     * used to add measurements in this context. Then context is changed to next one
+     * and so forth.
+     * 
+     * @return the {@link Tags} abstraction to use to build the context for this
+     *         cycle.
+     */
+    Tags openContext();
+
+    /**
+     * The reason for {@link Tags} API is to allow "describing" complex "assembled"
+     * multi-part keys to the {@link ProbingCycle} without the need to create
+     * intermediate objects like {@link String} concatenation would.
+     */
+    interface Tags {
```

Comment:
inner interface. Please pull it into top level.

Reply:
In this case I am really against extracting `Tags` from `ProbeCycle`. Its not a stand alone API. It is part of the `ProbeCycle` API that just has been isolated to force usage in a certain pattern and to limit access to particular methods when passing the cycle to `Tagging` that only should be able to use the methods found on `Tags`. Implementation wise this is the same class implementing it and  splitting this up does not make sense.

Reply:
I'm not a fan of internal interfaces. But in this case I can live with it.